src/workerd/util/sqlite-test.c++

```diff
@@ -926,5 +926,152 @@ KJ_TEST("SQLite failed statement reset") {
   KJ_EXPECT(db.run("SELECT COUNT(*) FROM things").getInt(0) == 4);
 }
 
+class MockRollbackCallback {
+public:
+  kj::Function<void()> create() {
+    KJ_ASSERT(!created);
+    created = true;
+    return [this, destructor = kj::defer([this]() { destroyed = true; })]() {
+      KJ_ASSERT(!called, "callback called multiple times?");
+      called = true;
+    };
+  }
+
+  bool isStillLive() {
+    return !destroyed && !called;
+  }
+  bool wasRolledBack() {
+    return called && destroyed;
+  }
+  bool wasCommitted() {
+    return !called && destroyed;
+  }
+
+private:
+  bool created = false;
+  bool called = false;
+  bool destroyed = false;
+};
+
+KJ_TEST("SQLite onRollback") {
+  auto dir = kj::newInMemoryDirectory(kj::nullClock());
+  SqliteDatabase::Vfs vfs(*dir);
+  SqliteDatabase db(vfs, kj::Path({"foo"}), kj::WriteMode::CREATE | kj::WriteMode::MODIFY);
+
+  // With no transactions open, the callback is dropped immediately.
+  {
+    MockRollbackCallback cb;
+    db.onRollback(cb.create());
+    KJ_EXPECT(cb.wasCommitted());
+  }
+
+  // Committed transactions drop the callback without invoking it.
+  {
+    db.run("BEGIN TRANSACTION");
+
+    MockRollbackCallback cb;
+    db.onRollback(cb.create());
+    KJ_EXPECT(cb.isStillLive());
+
+    db.run("COMMIT TRANSACTION");
+
+    KJ_EXPECT(cb.wasCommitted());
+  }
+
+  {
+    db.run("SAVEPOINT foo");
+
+    MockRollbackCallback cb;
+    db.onRollback(cb.create());
+    KJ_EXPECT(cb.isStillLive());
+
+    db.run("RELEASE SAVEPOINT foo");
+
+    KJ_EXPECT(cb.wasCommitted());
+  }
+
+  // Rollbacks invoke the callback.
+  {
+    db.run("BEGIN TRANSACTION");
+
+    MockRollbackCallback cb;
+    db.onRollback(cb.create());
+    KJ_EXPECT(cb.isStillLive());
+
+    db.run("ROLLBACK TRANSACTION");
+
+    KJ_EXPECT(cb.wasRolledBack());
+  }
+
+  {
+    db.run("SAVEPOINT foo");
+
+    MockRollbackCallback cb;
+    db.onRollback(cb.create());
+    KJ_EXPECT(cb.isStillLive());
+
+    db.run("ROLLBACK TO SAVEPOINT foo");
+    KJ_EXPECT(cb.wasRolledBack());
+
+    // The savepoint still exists until we release it...
+    db.run("RELEASE SAVEPOINT foo");
+  }
+
+  // Prepared statements work.
+  {
+    auto begin = db.prepare("BEGIN TRANSACTION");
+    auto commit = db.prepare("COMMIT TRANSACTION");
+
+    MockRollbackCallback cb1;
```

Comment:
nit: might be worth leaving a "// No transactions are open yet so the callback is dropped immediately." comment so folks aren't confused by `cb1.wasCommitted()` is true.